File: clients/swift/Sources/AresTransport/Orchestrator.swift
import Foundation

/// Connect N participants serially (clear per-participant failures); close all.
public enum Orchestrator {
    public static func connectAll(serverURL: String, pseudonyms: [String], sessionID: String,
                                  authSecret: String, timeout: TimeInterval = 60) async throws -> [Session] {
        var sessions: [Session] = []
        for p in pseudonyms {
            sessions.append(try await Session.connect(serverURL: serverURL, pseudonym: p,
                sessionID: sessionID, authSecret: authSecret, defaultTimeout: timeout))
        }
        return sessions
    }

    public static func closeAll(_ sessions: [Session]) async {
        for s in sessions { await s.close() }
    }
}

File: clients/kotlin/ares-client/src/main/kotlin/ares/client/transport/Orchestrator.kt
// SPDX-License-Identifier: Apache-2.0

package ares.client.transport

/**
 * Utility that connects and disconnects multiple [Session] instances.
 *
 * Mirrors Swift [Orchestrator] semantics. Sessions are opened serially so that
 * individual failures are attributed to their pseudonym.
 */
object Orchestrator {
    /**
     * Open one [Session] per pseudonym in [pseudonyms], all joining [sessionID].
     *
     * @param serverURL   base server URL passed to [Session.connect]
     * @param pseudonyms  ordered list of participant pseudonyms
     * @param sessionID   ARES session identifier shared by all participants
     * @param authSecret  shared HMAC secret; empty string disables auth
     * @return list of connected sessions in the same order as [pseudonyms]
     */
    fun connectAll(
        serverURL: String,
        pseudonyms: List<String>,
        sessionID: String,
        authSecret: String,
    ): List<Session> = pseudonyms.map { Session.connect(serverURL, it, sessionID, authSecret) }

    /**
     * Close all sessions, swallowing individual errors so that a single failure
     * does not prevent the remaining sessions from being cleaned up.
     */
    fun closeAll(sessions: List<Session>) {
        sessions.forEach { runCatching { it.close() } }
    }
}
